### PR TITLE
8313956: focusWithin on parents of a newly-added focused node is not updated

### PR DESCRIPTION
**Pull request by mstr2**
This PR fixes an issue with the way `focusWithin` bits are adjusted in the scene graph. Previously, the `focusWithin` counts of all parents of a removed node would be decreased if the removed node has a non-zero `focusWithin` count. This PR adds logic for the opposite scenario: the `focusWithin` count is increased on all parents of a newly-added node that is already focused (or contains other focused nodes).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8313956](https://bugs.openjdk.org/browse/JDK-8313956): focusWithin on parents of a newly-added focused node is not updated (**Bug** - P4)

### Reviewers
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**) 🔄 Re-review required (review applies to [fd85c902](https://git.openjdk.org/jfx/pull/1210/files/fd85c90268bafaaaa8371d7dae9f54d228892810))
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1210/head:pull/1210` \
`$ git checkout pull/1210`

Update a local copy of the PR: \
`$ git checkout pull/1210` \
`$ git pull https://git.openjdk.org/jfx.git pull/1210/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1210`

View PR using the GUI difftool: \
`$ git pr show -t 1210`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1210.diff">https://git.openjdk.org/jfx/pull/1210.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1210#issuecomment-1679654885)

### CODE REVIEW

**Comment by mstr2** [comment]
@Maran23 maybe you could take a look, too.

**Comment by Maran23** [comment]
> @Maran23 maybe you could take a look, too.

Yes! It is on my list.

**Comment by arapte** [comment]
Looks all neat to me. The FocusTest does cover a variety of scenarios.
The change is specific to FocusWithin and seems unlikely to have any side effects.

**Review by arapte — APPROVED** [review]

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java [hunk 1]
```diff
@@ -8171,22 +8174,38 @@ final void notifyFocusListeners() {
...
+        }
+
+        node = newParent;
+        while (node != null) {
+            node.focusWithin.adjust(focusWithin.count);
+            node = node.getParent();
+        };
+
+        // Since focus changes are atomic, we only fire change notifications after
+        // all changes are committed on all old and new parents.
```
Comment (hjohn) [inline modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8201]: Is it possible a user can interfere with this process by adding a listener to `focusWithinProperty`?  It fires events when it goes from `true` to `false` and vice versa.

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8201]: I also noticed this code, which I think will not do the correct thing when `change` is not 1 or -1:

        void adjust(int change) {
            count += change;

            if (count == 1) {
                set(true);
            } else if (count == 0) {
                set(false);
            }
        }

For example, if focus within currently is `false`, and it is "adjusted" by 5, the count goes from 0 to 5 in one go.  It will remain `false` then.  Is this intended?

**Review by hjohn — COMMENTED** [review]
I had a look, but have some questions.

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8201]: > Is it possible a user can interfere with this process by adding a listener to `focusWithinProperty`? It fires events when it goes from `true` to `false` and vice versa.

Can disregard this, it seems the property uses a special base class that breaks with the general property contract.

**Comment by Maran23** [comment]
One thing I noticed in the example in the ticket is, that when I click the button below the table, the `focusWithin` is still 1.
image

It seems like the `focusWithin` count is always 2 when the `TableView` is selected. When clicking the button, it will go down to 1, but not 0.

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java (hunk 1 above)
Comment (mstr2) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8201]: > I also noticed this code, which I think will not do the correct thing when `change` is not 1 or -1:
> 
> ```
>     void adjust(int change) {
>         count += change;
> 
>         if (count == 1) {
>             set(true);
>         } else if (count == 0) {
>             set(false);
>         }
>     }
> ```
> 
> For example, if focus within currently is `false`, and it is "adjusted" by 5, the count goes from 0 to 5 in one go. It will remain `false` then. Is this intended?

Good catch! I've added another test for this edge case.

**Comment by mstr2** [comment]
@arapte I've added another test for an edge case, can you review again?

**Comment by Maran23** [comment]
> That's because the table row remains `focused` even if you change the scene's focus owner to the "focus lost" button

I see. But to be honest, the table focus code is not optimal as it contains multiple focus workarounds which can be better solved with `focusWithin`. So we should first look what can be improved here. Thanks for looking that up.

One a side note, I remember that there is a ticket to clarify what to do when the `focusWithin` count is negative. We don't need to solve this in this PR, bur maybe for the future we should consider logging an info or something similar, in case we find another case where this can get negative.

**Comment by mstr2** [comment]
> One a side note, I remember that there is a ticket to clarify what to do when the `focusWithin` count is negative. We don't need to solve this in this PR, bur maybe for the future we should consider logging an info or something similar, in case we find another case where this can get negative.

The `focusWithin` count cannot be negative, and there is no possible way for applications to misuse the API to cause a negative count. If it ends up being negative, that's because of a bug, and I think that bugs should be fixed and not logged.

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java [hunk 2]
```diff
@@ -8323,7 +8323,7 @@ public String getName() {
         void adjust(int change) {
             count += change;
 
-            if (count == 1) {
+            if (change > 0) {
                 set(true);
             } else if (count == 0) {
                 set(false);
```
Comment (hjohn) [inline modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8329]: This looks a bit odd due to the use of two different variables, but it works (`count > 0` would also have worked).  I personally would favor something without any if statements as it leaves less room for misinterpretation in this case:

     count += change;

     set(count > 0);

File: modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java [hunk 3]
```diff
@@ -1128,4 +1128,35 @@ class N extends Group {
...
+        N node1, node2, node3, node4;
+
+        scene.setRoot(
+            node1 = new N(
+                node2 = new N()
+            ));
+
+        node3 = new N(
+            node4 = new N());
+
```
Comment (hjohn) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java:1146]: ignore: I think this looks a bit inconsistent (in the first version not all parenthesis are closed on the last line, while they are in the second version).  Personally I favor this, similar to how curly braces are used):
```suggestion
        scene.setRoot(
            node1 = new N(
                node2 = new N()
            )
        );

        node3 = new N(
            node4 = new N()
        );

```

**Review by hjohn — APPROVED** [review]
LGTM

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java (hunk 2 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8329]: set(count > 0);
looks so much better

Comment (mstr2) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java]: Agreed.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java (hunk 3 above)
Comment (mstr2) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java]: Done.

**Comment by andy-goryachev-oracle** [comment]
question: what happens when the user navigates away when editing the cell?  it looks like the fix correctly handles the situation.  are there any other possible scenarios?

Screenshot 2023-08-18 at 13 53 02

**Review by hjohn — APPROVED** [review]

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java [hunk 4]
```diff
@@ -8171,22 +8174,38 @@ final void notifyFocusListeners() {
...
-                node = node.getParent();
-            }
+    private void updateParentsFocusWithin(Node oldParent, Node newParent) {
+        if (!focusWithin.get()) {
+            return;
+        }
+
+        Node node = oldParent;
+        while (node != null) {
+            node.focusWithin.adjust(-focusWithin.count);
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8190]: suppose we are removing a top level Pane with a bunch of Nodes in it that have non-zero focusWithin.count values.
the parent of said Pane gets a zero count (which is correct), whereas the children of said Pane have their counts unchanged (I assume).

what happens when the same Pane gets added back to the scene graph?  wouldn't we enter an inconsistent state?

Comment (mstr2) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8190]: > suppose we are removing a Pane with a bunch of Nodes in it that have non-zero focusWithin.count values. the parent of said Pane gets a zero count (which is correct), whereas the children of said Pane have their counts unchanged (I assume).

Correct.

> what happens when the same Pane gets added back to the scene graph? wouldn't we enter an inconsistent state?

That's what this patch fixes. If we later add the pane that contains a number of focused nodes back to the scene graph, the `focusWithin` count of its new parents (which was 0 before) is now incremented to match the number of contained focused nodes. The code starts in L8194.

**Comment by mstr2** [comment]
> question: what happens when the user navigates away when editing the cell? it looks like the fix correctly handles the situation. are there any other possible scenarios?

I don't think there are, because now we're correctly handling the two scenarios that affect the `focusWithin` count:
1. Removing a non-zero focusWithin node from the scene graph
2. Adding a non-zero focusWithin node from the scene graph

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8190]: thank you for explanations!

one more question: suppose, prior to removal of the Pane it contained a focused (Scene.focusOwner) Node.
1. once the Pane is removed, what happens to Scene.focusOwner?
2. once the Pane is added back, would we have an inconsistent state in respect to Scene.focusOwner (if some other node requested focus in the mean time, wouldn't we have two focused nodes?)

Comment (mstr2) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8190]: > thank you for explanations!
> 
> one more question: suppose, prior to removal of the Pane it contained a focused (Scene.focusOwner) Node.
> 
> 1. once the Pane is removed, what happens to Scene.focusOwner?

There's quite a bit of logic happening. `Scene.focusCleanup()` is a good starting point. Basically, we need to find a new focus owner if the old focus owner is removed from the scene graph.

> 2. once the Pane is added back, would we have an inconsistent state in respect to Scene.focusOwner (if some other node requested focus in the mean time, wouldn't we have two focused nodes?)

It's not inconsistent, but it wouldn't be the same state as before. The `focused` state would remain, but `Scene.focusOwner` would not be restored to the previous node. That part of the focus story hasn't changed with the introduction of `focus-visible` and `focus-within`.

**Comment by Maran23** [comment]
> The `focusWithin` count cannot be negative, and there is no possible way for applications to misuse the API to cause a negative count. If it ends up being negative, that's because of a bug, and I think that bugs should be fixed and not logged.

Yes, the reason I was asking this is that we actually have this problem right now, that it could be negative, although it should not. If we would have logged that, we probably found that bug much earlier. In any case, I think with the new logic and the `count > 0`, it is fine as is.

**Review by Maran23 — APPROVED** [review]

File: modules/javafx.graphics/src/main/java/javafx/scene/Node.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/javafx/scene/Node.java:8190]: So, in the scenario # 2, we would arrive at situation when the previously focused Node is still decorated as having the focus, but won't be getting the events since Scene.focusOwner is different?  I suppose it's the way the whole thing is designed to work.
Thank you for the explanations!

**Review by andy-goryachev-oracle — APPROVED** [review]

File: modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java (hunk 3 above)
Comment (tcfurrer) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/FocusTest.java]: FYI: Using jfx21, I have observed a case where the count becomes negative. If I can create a minimal example, I will file a bug report.  However it's happening in a complex application, so it make take time for me to reduce it down to something I can reasonably provide.

**Comment by tcfurrer** [comment]
> > The `focusWithin` count cannot be negative, and there is no possible way for applications to misuse the API to cause a negative count. If it ends up being negative, that's because of a bug, and I think that bugs should be fixed and not logged.
> 
> Yes, the reason I was asking this is that we actually have this problem right now, that it could be negative, although it should not. If we would have logged that, we probably found that bug much earlier. In any case, I think with the new logic and the `count > 0`, it is fine as is.

Just to clarify...
If I have a case (with jfx21) where the count goes negative, is this still a bug that I should file?  Or are you saying that it's understood as expected now?

**Comment by tcfurrer** [comment]
@mstr2 or @Maran23 Is the term "focus owner" equivalent to "Node that currently has input focus"?

I ask because the javadoc on Node.focusWithin says "Indicates whether this Node or any of its descendants currently has the input focus.", but in discussion comments it has been said that focusWithin includes a TableView's "focused item" even if that is not the focus owner.

It would be great if the javadoc for Node.focusWithin could be improved to avoid this kind of widespread confusion about what it really means...

**Comment by mstr2** [comment]
> > > The `focusWithin` count cannot be negative, and there is no possible way for applications to misuse the API to cause a negative count. If it ends up being negative, that's because of a bug, and I think that bugs should be fixed and not logged.
> > 
> > 
> > Yes, the reason I was asking this is that we actually have this problem right now, that it could be negative, although it should not. If we would have logged that, we probably found that bug much earlier. In any case, I think with the new logic and the `count > 0`, it is fine as is.
> 
> @Maran23 Just to clarify... If I have a case (with jfx21) where the count goes negative, is this still a bug that I should file? Or are you saying that it's understood as expected behavior now?

If you have a situation where the focusWithin count is negative, please file it as a bug. That's not expected behavior.

**Comment by tcfurrer** [comment]
@Maran23 Ok, I will try to get a bug filed eventually.  It will take time because I've only observed this in a complex application, and it may be difficult to produce a minimal example.

**Comment by mstr2** [comment]
> @mstr2 or @Maran23 Does "focus owner" == "input focus" == "scene focus owner" == where key events will be sent?

This is true in general. Note that an unfocused window will not receive input events, but the scene can have a focus owner nonetheless.

> I ask because the javadoc on Node.focusWithin says "Indicates whether this Node or any of its descendants currently has the **_input focus_**.", but in discussions it has been said that focusWithin includes a TableView's "focused item" even if that is not the scene's current focus owner (i.e. where key events will be sent).
> 
> It would be great if the javadoc for Node.focusWithin could be improved to avoid this kind of widespread confusion about what it really means.

Yes, the documentation is misleading. We should at the very least fix that.

By the way, there's an ongoing discussion on the OpenJFX mailing list: https://mail.openjdk.org/pipermail/openjfx-dev/2023-October/042864.html
